src/node/jsonrpc.rs

```diff
@@ -0,0 +1,149 @@
+use crate::node::config::{
+    IPCAgentJsonRPCNodeConfig, DEFAULT_JSON_RPC_ENDPOINT, DEFAULT_JSON_RPC_VERSION,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::convert::Infallible;
+use warp::Filter;
+
+/// The IPC JSON RPC node that contains all the methods and handlers. The underlying implementation
+/// is using `warp`.
+///
+/// Note that currently only http json rpc is supported.
+///
+/// # Examples
+/// ```no_run
+/// use agent::node::config::IPCAgentJsonRPCNodeConfig;
+/// use agent::node::jsonrpc::IPCAgentJsonRPCNode;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let n = IPCAgentJsonRPCNode::new(IPCAgentJsonRPCNodeConfig::default());
+///     n.run().await;
+/// }
+/// ```
+pub struct IPCAgentJsonRPCNode {
+    config: IPCAgentJsonRPCNodeConfig,
+}
+
+impl IPCAgentJsonRPCNode {
+    pub fn new(config: IPCAgentJsonRPCNodeConfig) -> Self {
+        Self { config }
+    }
+
+    /// Runs the node in the current thread
+    pub async fn run(&self) {
+        log::info!("rpc node started at {:?}", self.config.addr());
+        warp::serve(json_rpc_filter()).run(self.config.addr()).await;
+    }
+}
+
+// Internal implementations
+
+/// Create the json_rpc filter
+fn json_rpc_filter() -> impl Filter<Extract = (warp::reply::Json,), Error = warp::Rejection> + Copy
+{
+    warp::post()
+        .and(warp::path(DEFAULT_JSON_RPC_ENDPOINT))
+        .and(warp::body::bytes())
+        .and_then(process)
+}
+
+/// The process method for the JSON RPC node.
+async fn process(bytes: bytes::Bytes) -> Result<warp::reply::Json, Infallible> {
+    log::debug!("received bytes = {:?}", bytes);
+
+    match serde_json::from_slice::<JSONRPCParam>(bytes.as_ref()) {
+        Ok(p) => {
+            let JSONRPCParam {
+                id,
+                method,
+                params,
+                jsonrpc,
+            } = p;
+
+            log::debug!("received method = {method:?} and params = {params:?}");
+
+            Ok(warp::reply::json(&JSONRPCResponse {
+                id,
+                jsonrpc,
+                result: (),
+            }))
+        }
+        Err(e) => {
+            log::error!("cannot parse parameter due to {e:?}");
+            Ok(warp::reply::json(&JSONRPCResponse {
+                id: 0,
+                jsonrpc: String::from(DEFAULT_JSON_RPC_VERSION),
+                result: serde_json::Value::String(String::from("Cannot parse parameters")),
+            }))
+        }
+    }
+}
+
+/// Follows: https://ethereum.org/en/developers/docs/apis/json-rpc/#curl-examples
+#[derive(Serialize, Deserialize)]
+struct JSONRPCParam {
+    pub id: u16,
+    pub jsonrpc: String,
+    pub method: String,
+    pub params: Value,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct JSONRPCResponse<T: Serialize> {
+    pub id: u16,
+    pub jsonrpc: String,
+    pub result: T,
+}
```

Comment:
Can we extract these to a `types.rs` mod?

Reply:
reason I put it here because in the scope of the current PR, there is no other struct or module using this. It's completely internal to this file. Maybe I will move it to a separate file when there is a use case for this?

Reply:
Gotcha, so they are not publicly exposed yet. I was expecting them to be in the future once the CLI and external libraries need these types to perform jsonrpc requests. I am fine leaving them here for now then, nonetheless can you move them then to the beginning of the file, before the functions declarations so we have first the type and trait definitions and then the functions?

Reply:
updated and resolving this.